doc/doxygen/src/build-system-cpu-conventions.md

```diff
@@ -0,0 +1,180 @@
+================================================
+
+[TOC]
+
+Introduction
+============
+
+This page describes some conventions for defining CPU in the build system.
+Not all of them are currently listed but only particular ones.
+
+
+`CPU` defining multiple `CPU_MODEL`s         {#cpu-defining-multiple-cpu-models}
+====================================
+
+Context                                                               {#context}
+-------
+
+With the increasing number of supported CPU models which have lots of
+similarities with other ones from the same family, there is a need to declare
+them in a single directory but still allow model specific definitions
+and behaviors. This leads to the difference between `CPU` and `CPU_MODEL`.
+
+Several handling were implemented in the build system using shell invocations
+and algorithmic resolution based on the models characters.
+However, due to the algorithmic resolution one cannot know anymore the value of
+the specific CPU model definitions without executing the build system and the
+shell `sed` magic.
+
+This causes both issue to read the definition and slow execution time for
+definitions that are a static mapping from the exact CPU model.
+
+With the way of defining boards and CPU, some informations should be defined
+either for the description, dependency resolution or compilation
+(respectively in the `Makefile.features`, the `Makefile.dep` or `Makefile.include`)
+so may want to define only each piece of information required for each step.
+
+
+Different solutions                                       {#different-solutions}
+-------------------
+
+This describes different possible solutions, and the cases on which each of them applies.
+
+If only one should be used, then it should be solution 1 to go toward
+declarative syntax and support all cases.
+But due to the overhead, intermediate single file declarative solution 3 and 4
+are good candidates too.
+
+As currently dependencies are not declarative, only solutions 1 and 2 would be
+able to declare specific `CPU_MODEL` dependencies. Solutions 3 and 4 are
+declarative only.
+
+The main goal is to remove the algorithmic and shell based resolutions for this
+if possible.
+
+
+### 1. File based handling                                {#file-based-handling}
+
+Somehow, having all `CPU_MODEL` in one directory is just to save common
+parts and keep the `cpu` directory tidy, all the different ones could still
+be defined in their own directory, but nested in the common one.
+
+This could transfer to have for each `CPU_MODEL` for a `CPU` the following files:
+
+~~~~~~~~~~~~~~~~~~~ {.mk}
+$(RIOTCPU)/$(CPU)/models/$(CPU_MODEL)/Makefile.features
+$(RIOTCPU)/$(CPU)/models/$(CPU_MODEL)/Makefile.dep
+$(RIOTCPU)/$(CPU)/models/$(CPU_MODEL)/Makefile.include
+$(RIOTCPU)/$(CPU)/models/$(CPU_MODEL)/doc.txt
+~~~~~~~~~~~~~~~~~~~
+
+And then each main `$(CPU)` file will include the `$(CPU_MODEL)` specific one.
```

Comment:
I believe I misunderstood this part when I first read it. I think this should be the other way around. A board should specify which `CPU_MODEL` it's using, then the model is in charge of including (or not) any common files like the `CPU` (which may then include other common, and thus more generic, files).

Files should be included from more specific to less specific. `CPU_MODEL` depends on `BOARD`, and `CPU` depends on `CPU_MODEL`, not the other way around.

Reply:
yes I agree.